File: datasphere/ml/validation.py
"""Dataset validation — schema checks without modifying raw data."""

from __future__ import annotations

from dataclasses import dataclass, field

import pandas as pd

from datasphere.ml.config import EXCLUDED_FEATURE_COLUMNS, RISK_LEVELS, TARGET_COLUMN

REQUIRED_COLUMNS: tuple[str, ...] = (
    "student_id",
    "record_generated_date",
    "enrollment_date",
    "academic_year",
    "country",
    "state_province",
    "location_type",
    "age",
    "gender",
    "grade_level",
    "school_type",
    "distance_to_school_km",
    "mode_of_transport",
    "household_income_monthly_usd",
    "mother_education_level",
    "father_education_level",
    "family_size",
    "number_of_siblings",
    "is_orphan",
    "single_parent_household",
    "household_has_electricity",
    "household_has_internet_access",
    "owns_smartphone_or_computer",
    "child_labor_involvement",
    "attendance_rate_pct",
    "average_test_score_pct",
    "previous_year_pass",
    "number_of_school_transfers",
    "disciplinary_incidents_count",
    "bullying_incidents_reported",
    "health_issues_reported",
    "special_needs_status",
    "free_meal_program_enrolled",
    "scholarship_received",
    "extracurricular_participation",
    "teacher_student_ratio",
    "school_infrastructure_score",
    "early_marriage_risk_flag",
    "community_conflict_zone",
    "seasonal_migration_family",
    "ngo_intervention_present",
    "literacy_program_enrolled",
    "language_barrier_flag",
    "report_source",
    "case_notes",
    "dropout_probability_score",
    TARGET_COLUMN,
)


@dataclass
class ValidationReport:
    passed: bool
    row_count: int
    column_count: int
    missing_columns: list[str] = field(default_factory=list)
    unexpected_columns: list[str] = field(default_factory=list)
    duplicate_student_ids: int = 0
    invalid_target_values: list[str] = field(default_factory=list)
    messages: list[str] = field(default_factory=list)


def validate_raw_dataset(df: pd.DataFrame) -> ValidationReport:
    """Validate schema and basic integrity of the raw dataset."""
    missing = [col for col in REQUIRED_COLUMNS if col not in df.columns]
    unexpected = [col for col in df.columns if col not in REQUIRED_COLUMNS]
    duplicate_ids = int(df["student_id"].duplicated().sum()) if "student_id" in df.columns else 0

    invalid_targets: list[str] = []
    if TARGET_COLUMN in df.columns:
        values = set(df[TARGET_COLUMN].dropna().astype(str).unique())
        invalid_targets = sorted(values - set(RISK_LEVELS))

    messages: list[str] = []
    if missing:
        messages.append(f"Missing required columns: {missing}")
    if duplicate_ids:
        messages.append(f"Duplicate student_id count: {duplicate_ids}")
    if invalid_targets:
        messages.append(f"Invalid target values: {invalid_targets}")

    passed = not missing and not invalid_targets
    return ValidationReport(
        passed=passed,
        row_count=len(df),
        column_count=len(df.columns),
        missing_columns=missing,
        unexpected_columns=unexpected,
        duplicate_student_ids=duplicate_ids,
        invalid_target_values=invalid_targets,
        messages=messages,
    )


def assert_no_leakage_columns(feature_columns: list[str]) -> None:
    """Raise if any leakage or excluded column appears in the feature matrix."""
    forbidden = set(EXCLUDED_FEATURE_COLUMNS) - {TARGET_COLUMN}
    present = forbidden.intersection(feature_columns)
    if present:
        raise ValueError(f"Leakage or excluded columns found in X: {sorted(present)}")

File: frontend/src/utils/validation.ts
import type { ModelSchema, StudentInput } from '../types/student'

export function createDefaultStudent(schema?: ModelSchema): StudentInput {
  const opts = schema?.categorical_options ?? {}
  const pick = (key: string, fallback: string) => opts[key]?.[0] ?? fallback
  const today = new Date().toISOString().slice(0, 10)
  return {
    record_generated_date: today,
    enrollment_date: '2023-09-01',
    academic_year: pick('academic_year', '2023-2024'),
    country: pick('country', 'Kenya'),
    state_province: pick('state_province', 'Nairobi'),
    location_type: pick('location_type', 'Urban'),
    age: 14,
    gender: pick('gender', 'Female'),
    grade_level: 8,
    school_type: pick('school_type', 'Government'),
    distance_to_school_km: 2.5,
    mode_of_transport: pick('mode_of_transport', 'Walking'),
    household_income_monthly_usd: 250,
    mother_education_level: pick('mother_education_level', 'Secondary'),
    father_education_level: pick('father_education_level', 'Primary'),
    family_size: 5,
    number_of_siblings: 2,
    is_orphan: false,
    single_parent_household: false,
    household_has_electricity: true,
    household_has_internet_access: false,
    owns_smartphone_or_computer: false,
    child_labor_involvement: false,
    attendance_rate_pct: 85,
    average_test_score_pct: 65,
    previous_year_pass: true,
    number_of_school_transfers: 0,
    disciplinary_incidents_count: 0,
    bullying_incidents_reported: false,
    health_issues_reported: false,
    special_needs_status: false,
    free_meal_program_enrolled: true,
    scholarship_received: false,
    extracurricular_participation: true,
    teacher_student_ratio: 35,
    school_infrastructure_score: 60,
    early_marriage_risk_flag: false,
    community_conflict_zone: false,
    seasonal_migration_family: false,
    ngo_intervention_present: false,
    literacy_program_enrolled: false,
    language_barrier_flag: false,
    report_source: pick('report_source', 'School Record'),
  }
}

export function validateStudent(
  student: StudentInput,
  schema: ModelSchema,
): Partial<Record<keyof StudentInput, string>> {
  const errors: Partial<Record<keyof StudentInput, string>> = {}
  for (const [key, [min, max]] of Object.entries(schema.numeric_bounds)) {
    const field = key as keyof StudentInput
    const value = student[field]
    if (typeof value === 'number' && (value < min || value > max)) {
      errors[field] = `Must be between ${min} and ${max}`
    }
  }
  for (const [key, allowed] of Object.entries(schema.categorical_options)) {
    const field = key as keyof StudentInput
    const value = student[field]
    if (typeof value === 'string' && !allowed.includes(value)) {
      errors[field] = 'Invalid selection'
    }
  }
  return errors
}
